readthedocs/vcs_support/backends/git.py

```diff
@@ -27,22 +27,15 @@ class Backend(BaseVCS):
     fallback_branch = "master"  # default branch
     repo_depth = 50
 
-    def __init__(self, *args, **kwargs):
+    def __init__(self, *args, use_token=False, **kwargs):
         super().__init__(*args, **kwargs)
-        self.token = kwargs.get("token")
+        self.use_token = use_token
         self.repo_url = self._get_clone_url()
 
     def _get_clone_url(self):
-        if "://" in self.repo_url:
-            hacked_url = self.repo_url.split("://")[1]
-            hacked_url = re.sub(".git$", "", hacked_url)
-            clone_url = "https://%s" % hacked_url
-            if self.token:
-                clone_url = "https://{}@{}".format(self.token, hacked_url)
-                return clone_url
-            # Don't edit URL because all hosts aren't the same
-            # else:
-            #     clone_url = 'git://%s' % (hacked_url)
+        if self.repo_url.startswith(("https://", "http://")) and self.use_token:
+            parsed_url = urlparse(self.repo_url)
+            return f"{parsed_url.scheme}://$READTHEDOCS_GIT_CLONE_TOKEN@{parsed_url.netloc}{parsed_url.path}"
```

Comment:
Note that we will need to hide/obfuscate this somehow so it's not shown in the build details page.

Reply:
It's an environment variable, isn't shown in the buil

Reply:
Be careful with this. We are using the same pattern in https://github.com/readthedocs/readthedocs.org/blob/main/dockerfiles/Dockerfile#L79 and it's shown when the command fails for any reason.

Reply:
That's probably Docker's behavior to show the failed command? I did a test with an invalid token, token without access, and invalid repo, and git never reveals the token in the error message.

Reply:
> I think we should probably show a "fake" command (without the <token>@ part) here to avoid potential leaks here.

Any potential leak would be in the output of the command, not in the command itself. The command is already saved as `$READTHEDOCS_GIT_CLONE_TOKEN`, the token is never saved, so it's never shown.

Reply:
Great!